provider/oci/environ.go

```diff
@@ -123,54 +347,344 @@ func (e *Environ) SetConfig(cfg *config.Config) error {
+	assignPublicIp := true
+	bootSource := ociCore.InstanceSourceViaImageDetails{
+		ImageId:             &image,
+		BootVolumeSizeInGBs: &rootDiskSize,
+	}
+	instanceDetails := ociCore.LaunchInstanceDetails{
+		AvailabilityDomain: &zone,
+		CompartmentId:      e.ecfg().compartmentID(),
+		SourceDetails:      bootSource,
+		Shape:              &spec.InstanceType.Name,
+		CreateVnicDetails: &ociCore.CreateVnicDetails{
+			SubnetId:       network.Id,
+			AssignPublicIp: &assignPublicIp,
+			DisplayName:    &hostname,
+		},
+		DisplayName: &hostname,
+		Metadata: map[string]string{
+			"user_data": string(userData),
+		},
+		FreeformTags: tags,
+	}
+
+	request := ociCore.LaunchInstanceRequest{
+		LaunchInstanceDetails: instanceDetails,
+	}
+
+	response, err := e.Compute.LaunchInstance(context.Background(), request)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	instance, err := newInstance(response.Instance, e)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	machineId := response.Instance.Id
+	timeout := 10 * time.Minute
+	if err := instance.waitForMachineStatus(desiredStatus, timeout); err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Infof("started instance %q", *machineId)
+
+	if desiredStatus == ociCore.InstanceLifecycleStateRunning {
+		if err := instance.waitForPublicIP(ctx); err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	result := &environs.StartInstanceResult{
+		Instance: instance,
+		Hardware: instance.hardwareCharacteristics(),
+	}
+
+	return result, nil
 }
 
 // StopInstances implements environs.InstanceBroker.
-func (e *Environ) StopInstances(ctx context.ProviderCallContext, ids ...instance.Id) error {
-	return errors.NotImplementedf("StopInstances")
+func (e *Environ) StopInstances(ctx envcontext.ProviderCallContext, ids ...instance.Id) error {
+	ociInstances, err := e.getOciInstances(ids...)
+	if err == environs.ErrNoInstances {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	logger.Debugf("terminating instances %v", ids)
+	if err := e.terminateInstances(ociInstances...); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *Environ) terminateInstances(instances ...*ociInstance) error {
+	wg := sync.WaitGroup{}
+	wg.Add(len(instances))
+	errs := []error{}
+	instIds := []instance.Id{}
+	for _, oInst := range instances {
+		go func(inst *ociInstance) {
+			defer wg.Done()
+			if err := inst.deleteInstance(); err != nil {
+				instIds = append(instIds, inst.Id())
+				errs = append(errs, err)
+			} else {
+				err := inst.waitForMachineStatus(
+					ociCore.InstanceLifecycleStateTerminated,
+					5*time.Minute)
+				if err != nil && !errors.IsNotFound(err) {
+					instIds = append(instIds, inst.Id())
+					errs = append(errs, err)
+				}
+			}
+		}(oInst)
+	}
+	wg.Wait()
+	switch len(errs) {
+	case 0:
+		return nil
+	case 1:
+		return errors.Annotatef(errs[0], "failed to stop instance %s", instIds[0])
+	default:
+		return errors.Errorf(
```

Comment:
Indeed. Will add.

Reply:
I think it does work, because golang has a "if getting a format specifier, and the value being based is a slice of objects of that type, then treat the format against each entry of the slice, and combine it into a list".

However, a  concrete test of this helps ensure that it really does work as we would like.

Reply:
added